models/repo_branch.go

```diff
@@ -52,6 +63,91 @@ func (repo *Repository) GetBranches() ([]*Branch, error) {
 	return GetBranchesByPath(repo.RepoPath())
 }
 
+// CreateNewBranch creates a new repository branch
+func (repo *Repository) CreateNewBranch(doer *User, oldBranchName, branchName string) (err error) {
+	repoWorkingPool.CheckIn(com.ToStr(repo.ID))
+	defer repoWorkingPool.CheckOut(com.ToStr(repo.ID))
+
+	localPath := repo.LocalCopyPath()
+
+	if err = discardLocalRepoBranchChanges(localPath, oldBranchName); err != nil {
+		return fmt.Errorf("discardLocalRepoChanges: %v", err)
+	} else if err = repo.UpdateLocalCopyBranch(oldBranchName); err != nil {
+		return fmt.Errorf("UpdateLocalCopyBranch: %v", err)
+	}
+
+	if err = repo.CheckoutNewBranch(oldBranchName, branchName); err != nil {
+		return fmt.Errorf("CreateNewBranch: %v", err)
+	}
+
+	if err = git.Push(localPath, git.PushOptions{
+		Remote: "origin",
+		Branch: branchName,
+	}); err != nil {
+		return fmt.Errorf("Push: %v", err)
+	}
+
+	return nil
+}
+
+// UpdateLocalCopyToCommit pulls latest changes of given commit from repoPath to localPath.
+// It creates a new clone if local copy does not exist.
+// This function checks out target commit by default, it is safe to assume subsequent
+// operations are operating against target commit when caller has confidence for no race condition.
+func UpdateLocalCopyToCommit(repoPath, localPath, commit string) error {
+	if !com.IsExist(localPath) {
```

Comment:
I believe this is duplicated code; this `if`/`else` can be replaced by
```
UpdateLocalCopyBranch(repoPath, localPath, repo.DefaultBranch)
```

Reply:
@ethantkoenig no it can not as they are different and calls methods in different order. Code from UpdateLocalCopyBranch will fail in some cases if not updated in correct order (at least for creating new branch from commit/tag)